Generated code for async functions in structs doesn't compile

### Describe the Bug

A clear and concise description of what the bug is.

### Steps to Reproduce

The issue is described in https://github.com/fabricedesre/wasm-cryptokey/issues/1, but here's a simpler test case:

Compile a basic wasm-bindgen project with:

```rust
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub struct AsyncTest;

#[wasm_bindgen]
impl AsyncTest {
    #[wasm_bindgen(constructor)]
    pub async fn new() -> Self {
        AsyncTest
    }

    #[wasm_bindgen]
    pub async fn call_me(&self) -> u32 {
        42
    }
}
```

### Expected Behavior

Code compiles properly

### Actual Behavior

Compilation error:
```shell
error[E0597]: `me` does not live long enough
 --> src/lib.rs:6:1
  |
6 | #[wasm_bindgen]
  | ^^^^^^^^^^^^^^-
  | |             |
  | |             `me` dropped here while still borrowed
  | borrowed value does not live long enough
  | argument requires that `me` is borrowed for `'static`
  |
  = note: this error originates in the attribute macro `wasm_bindgen::prelude::__wasm_bindgen_class_marker` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Additional Context

The generated code is:

```rust
pub async fn call_me(&self) -> u32
{
    #[automatically_derived] const
    __wasm_bindgen_generated_AsyncTest_call_me__const : () =
    {
        #[cfg_attr(all(target_arch = "wasm32", not(target_os = "emscripten")),
        export_name = "asynctest_call_me",)] pub extern "C" fn
        __wasm_bindgen_generated_AsyncTest_call_me(me : u32) -> <wasm_bindgen
        ::JsValue as wasm_bindgen::convert::ReturnWasmAbi> :: Abi
        {
            let _ret =
            {
                let me = unsafe
                {
                    <AsyncTest as wasm_bindgen::convert::RefFromWasmAbi>
                    ::ref_from_abi(me)
                } ; let me = & * me ; me.call_me()
            } ; <wasm_bindgen::JsValue as wasm_bindgen::convert::
            ReturnWasmAbi>::
            return_abi(wasm_bindgen_futures::
            future_to_promise(async move
            {
                <u32 as wasm_bindgen: __rt::IntoJsResult>::
                into_js_result(_ret.await)
            }).into())
        }
    } ; #[cfg(all(target_arch = "wasm32", not(target_os = "emscripten")))]
    #[automatically_derived] const __wbindgen_const_describe_asynctest_call_me
    : () =
    {
        #[no_mangle] #[doc(hidden)] pub extern "C" fn
        __wbindgen_describe_asynctest_call_me()
        {
            use wasm_bindgen :: describe :: * ; wasm_bindgen :: __rt ::
            link_mem_intrinsics() ; inform(FUNCTION) ; inform(0) ;
            inform(0u32) ; < wasm_bindgen :: JsValue as WasmDescribe > ::
            describe() ; < u32 as WasmDescribe > :: describe() ;
        }
    } ; #[cfg(target_arch = "wasm32")] #[automatically_derived]
    #[link_section = "__wasm_bindgen_unstable"] #[doc(hidden)] pub static
    __WASM_BINDGEN_GENERATED_dcebcd904eb6cb29 : [u8 ; 123usize] =
    {
        static _INCLUDED_FILES : & [& str] = & [] ; *
        b".\x00\x00\x00{\"schema_version\":\"0.2.82\",\"version\":\"0.2.82\"}E\x00\x00\x00\x01\x01\tAsyncTest\x00\x00\x00\x01\x07call_me\x01\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x1fwasm-cryptokey-11d3b729a3125508\x00"
    } ; 42
}
```

--- a/crates/backend/src/codegen.rs
+++ b/crates/backend/src/codegen.rs
@@ -470,9 +470,7 @@ impl TryToTokens for ast::Export {
                     quote! { () },
                     quote! { () },
                     quote! {
-                        wasm_bindgen_futures::spawn_local(async move {
-                            <#syn_ret as wasm_bindgen::__rt::Start>::start(#ret.await);
-                        })
+                        <#syn_ret as wasm_bindgen::__rt::Start>::start(#ret.await)
                     },
                 )
             } else {
@@ -480,9 +478,7 @@ impl TryToTokens for ast::Export {
                     quote! { wasm_bindgen::JsValue },
                     quote! { #syn_ret },
                     quote! {
-                        wasm_bindgen_futures::future_to_promise(async move {
-                            <#syn_ret as wasm_bindgen::__rt::IntoJsResult>::into_js_result(#ret.await)
-                        }).into()
+                        <#syn_ret as wasm_bindgen::__rt::IntoJsResult>::into_js_result(#ret.await)
                     },
                 )
             }
@@ -496,8 +492,32 @@ impl TryToTokens for ast::Export {
             (quote! { #syn_ret }, quote! { #syn_ret }, quote! { #ret })
         };
 
+        let mut call = quote! {
+            {
+                #(#arg_conversions)*
+                let #ret = #receiver(#(#converted_arguments),*);
+                #ret_expr
+            }
+        };
+
+        if self.function.r#async {
+            if self.start {
+                call = quote! {
+                    wasm_bindgen_futures::spawn_local(async move {
+                        #call
+                    })
+                }
+            } else {
+                call = quote! {
+                    wasm_bindgen_futures::future_to_promise(async move {
+                        #call
+                    }).into()
+                }
+            }
+        }
+
         let projection = quote! { <#ret_ty as wasm_bindgen::convert::ReturnWasmAbi> };
-        let convert_ret = quote! { #projection::return_abi(#ret_expr) };
+        let convert_ret = quote! { #projection::return_abi(#ret) };
         let describe_ret = quote! {
             <#ret_ty as WasmDescribe>::describe();
             <#inner_ret_ty as WasmDescribe>::describe();
@@ -523,13 +543,8 @@ impl TryToTokens for ast::Export {
                 )]
                 pub unsafe extern "C" fn #generated_name(#(#args),*) -> #projection::Abi {
                     #start_check
-                    // Scope all local variables to be destroyed after we call
-                    // the function to ensure that `#convert_ret`, if it panics,
-                    // doesn't leak anything.
-                    let #ret = {
-                        #(#arg_conversions)*
-                        #receiver(#(#converted_arguments),*)
-                    };
+
+                    let #ret = #call;
                     #convert_ret
                 }
             };

--- a/tests/wasm/futures.js
+++ b/tests/wasm/futures.js
@@ -15,6 +15,9 @@ exports.call_exports = async function() {
   await assert.rejects(wasm.async_throw_message(), /async message/);
   await assert.rejects(wasm.async_throw_jserror(), /async message/);
   await assert.rejects(wasm.async_throw_custom_error(), /custom error/);
+  assert.strictEqual("Hi, Jim!", await wasm.async_take_reference("Jim"));
+  const foo = await new wasm.AsyncStruct();
+  assert.strictEqual(42, await foo.method());
 };
 
 exports.call_promise = async function() {

--- a/tests/wasm/futures.rs
+++ b/tests/wasm/futures.rs
@@ -106,6 +106,26 @@ pub async fn async_throw_custom_error() -> Result<AsyncCustomReturn, AsyncCustom
     })
 }
 
+#[wasm_bindgen]
+pub async fn async_take_reference(x: &str) -> String {
+    format!("Hi, {x}!")
+}
+
+#[wasm_bindgen]
+pub struct AsyncStruct;
+
+#[wasm_bindgen]
+impl AsyncStruct {
+    #[wasm_bindgen(constructor)]
+    pub async fn new() -> AsyncStruct {
+        AsyncStruct
+    }
+
+    pub async fn method(&self) -> u32 {
+        42
+    }
+}
+
 #[wasm_bindgen_test]
 async fn test_promise() {
     assert_eq!(call_promise().await.as_string(), Some(String::from("ok")))

